fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java

```diff
@@ -756,78 +753,104 @@ public void upsertCommonTokenValues(List<ResourceTokenValueRec> values) {
         sortedTokenValues.sort(CommonTokenValue::compareTo);
 
-        // Build a string of parameter values we use in the query to drive the insert statement.
-        // The database needs to know the type when it parses the query, hence the slightly verbose CAST functions:
-        // VALUES ((CAST(? AS VARCHAR(1234)), CAST(? AS INT)), (...)) AS V(common_token_value, parameter_name_id, code_system_id)
-        StringBuilder inList = new StringBuilder(); // for the select query later
-        StringBuilder paramList = new StringBuilder();
-        for (int i=0; i<sortedTokenValues.size(); i++) {
-            if (paramList.length() > 0) {
-                paramList.append(", ");
-            }
-            paramList.append("(CAST(? AS VARCHAR(" + FhirSchemaConstants.MAX_TOKEN_VALUE_BYTES + "))");
-            paramList.append(",CAST(? AS INT))");
+        Optional<Integer> maxQuery = translator.maximumQueryParameters();
 
-            // also build the inList for the select statement later
-            if (inList.length() > 0) {
-                inList.append(",");
-            }
-            inList.append("(?,?)");
-        }
+        // Process segments up to maximum query parameter size.
+        for (int low = 0; low < sortedTokenValues.size();/*at end  of loop*/) {
+            int high = sortedTokenValues.size();
 
-        final String paramListStr = paramList.toString();
-        doCommonTokenValuesUpsert(paramListStr, sortedTokenValues);
+            if (!maxQuery.isEmpty() && high > maxQuery.get()/2) {
+                // 2 as there a pair inserted.
+                high = low + maxQuery.get()/2;
+            }
 
-        // Now grab the ids for the rows we just created. If we had a RETURNING implementation
-        // which worked reliably across all our database platforms, we wouldn't need this
-        // second query.
-        // Derby doesn't support IN LISTS with multiple members, so we have to join against
-        // a VALUES again. No big deal...probably similar amount of work for the database
-        StringBuilder select = new StringBuilder();
-        select.append("     SELECT ctv.code_system_id, ctv.token_value, ctv.common_token_value_id FROM ");
-        select.append("     (VALUES ").append(paramListStr).append(" ) AS v(token_value, code_system_id) ");
-        select.append("       JOIN common_token_values ctv ");
-        select.append("              ON ctv.token_value = v.token_value ");
-        select.append("             AND ctv.code_system_id = v.code_system_id ");
-
-        // Grab the ids
-        Map<CommonTokenValue, Long> idMap = new HashMap<>();
-        try (PreparedStatement ps = connection.prepareStatement(select.toString())) {
-            int a = 1;
-            for (CommonTokenValue tv: sortedTokenValues) {
-                ps.setString(a++, tv.getTokenValue());
-                ps.setInt(a++, tv.getCodeSystemId());
+            // Make sure we don't overflow.
```

Comment:
this if block can be inside the above if block, I think. It only applies of the high has been recalculated.